File: week6/sentimental-credit/credit.py
# TODO

from cs50 import get_int

# Initiates program
while True:
    # Prompts user to input card number
    card = get_int("Card: ")
    # Verifys number entered
    if card > 0:
        # Program proceeds if number entered.
        break



# Determine if the card number is valid, using Luhn's algorithm
# Create luhn_checksum variable
def luhn_checksum(card):
    # Iterates through the numbers
    def digits_of(n):
        # Returns digits as a string of numbers
        return [int(d) for d in str(n)]
    digits = digits_of(card)
    # Uses negative indexing to get list of odd digits
    odd_digits = digits[-1::-2]
    # Uses negative indexing to get list of even digits
    even_digits = digits[-2::-2]
    # Sets checksum equal to zero to begin computation
    checksum = 0
    # Adds all odd digit values
    checksum += sum(odd_digits)
    # Operation to multiply even digits by 2
    for d in even_digits:
        checksum += sum(digits_of(d*2))
    # Returns checksum module 10, to obtain last digit
    return checksum % 10



# Identify the type of credit card used
lenght = 0
visa = card
master = card
amex = card

# Differentiate between VISA, Master, AMEX
lenght = len(str(card))

# Identify if card is VISA
while visa >= 10:
    visa = int(visa/10)

# Identify if card is AMEX
while amex >= 10**13:
    amex = int(amex/10**13)

# Identify if card is Mastercard
while master >= 10**14:
    master = int(master/10**14)

# Print outcome of card type
if luhn_checksum(card) == 0:
    if visa == 4 and (lenght == 13 or lenght == 16):
        print("VISA")
    elif lenght == 15 and (amex == 34 or amex == 37):
        print("AMEX")
    elif lenght == 16 and (51 <= master <= 55):
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")

File: week1/credit/credit.c
// Tells compiler to load specified library into program.
#include <cs50.h>
// Tells compiler to load specified library into program.
#include <stdio.h>
// Tells compiler to load specified library into program.
#include <math.h>

// Prototype functions.
bool LuhnCheck(long CardNumber);
int GetCardLenght(long CardNumber);
void GetCardProvider(long CardNumber);


// This initiates code.
int main(void)
{
    // Prompts user to input card number.
    long CardNumber = get_long_long("Card Number to Validate: ");
    // Checks if card number is valid to Luhn's Algorithm, if not valid, prints INVALID.
    if (LuhnCheck(CardNumber) == true)
    {
        // Check if card is valid for a provider.
        GetCardProvider(CardNumber);

    }
    else
    {
        // Prints INVALID if number is not valid to Luhn's Algorithm.
        printf("INVALID\n");
    }
}

// Application of Luhn's Algorithm.
bool LuhnCheck(long CardNumber)
{
    // Sum that validates or not, the card.
    int CheckSum = 0;
    // Copy of card number.
    long tmp = CardNumber;

    // Iterates over each digit of the card number, checking for odds and evens.
    for (int pos = 0; tmp > 0; pos++)
    {
        if (pos % 2 == 0)
        {
            // Adds the last digit of the card to the sum.
            CheckSum += tmp % 10;
            // Excludes the last digit of the card
            tmp /= 10;
        }
        else
        {
            // For odd numbers it's necessary to *2
            int odd = (tmp % 10) * 2;
            // Makes sure we add each digit of the product and not the whole number.
            CheckSum += (odd / 10 + odd % 10);
            // Excludes the last digit of the card.
            tmp /= 10;
        }
    }
    // Returns true or false for the card validation. (If the sum's last digit is equal to zero).
    return (CheckSum % 10) == 0;
}

int GetCardLenght(long CardNumber)
{
    // Counts how many digits the card number has.
    int len = log10(CardNumber);
    return len + 1;
}

void GetCardProvider(long CardNumber)
{
    // Saves the card number lenght for use.
    int len = GetCardLenght(CardNumber);
    // It's going to store the first 2 digits of the card number.
    int prov;

    // Code Block for 16 digits lenght.
    if (len == 16)
    {
        // Uses scientific notation to reduce card number to 2 digits.
        prov = (CardNumber / 1e14);
        // Then checks said digits and prints it's provider.
        if (prov >= 51 && prov <= 55)
        {
            printf("MASTERCARD\n");
        }
        else if (prov >= 40 && prov <= 49)
        {
            printf("VISA\n");
        }
        // For check error.
        else
        {
            printf("INVALID\n");
        }
    }

    // Code block for 13 digits lenght.
    if (len == 13)
    {
        prov = (CardNumber / 1e12);
        if (prov == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    // Code block for 15 digits lenght.
    if (len == 15)
    {
        prov = (CardNumber / 1e13);
        if (prov == 34 || prov == 37)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    // Deals with any other lenght not specified by blocks above.
    else if (len <= 12 || len >= 17 || len == 14)
    {
        printf("INVALID\n");
    }
}